libnetwork/service_linux.go

```diff
@@ -264,39 +264,39 @@ var (
 	ingressMu       sync.Mutex // lock for operations on ingress
 	ingressProxyTbl = make(map[string]io.Closer)
 	portConfigMu    sync.Mutex
-	portConfigTbl   = make(map[PortConfig]int)
+	portConfigTbl   = make(map[*PortConfig]int)
```

Comment:
This changes the map keys from matching by-value (i.e. the `(name, protocol, target_port, published_port)` tuple) to matching by reference identity. `PortConfig{} == PortConfig{}` but `&PortConfig{} != &PortConfig{}`. The by-value behaviour is clearly the intended one, so this change will break things. And annoyingly [there isn't any guarantee that protobuf serialization is deterministic](https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510) so that wouldn't work either. While there may be tuning knobs in gogo protobuf to stop it from emitting the extra `XXX_` fields in `PortConfig`, that'd be only a stopgap measure as gogo protobuf is officially deprecated. Unfortunately, the only forward-compatible solution I can see is to define a parallel struct type by hand for use as map keys.